aries_vcx/src/protocols/connection/inviter/mod.rs

```diff
@@ -1,2 +1,235 @@
+        let response = Response::create()
+            .set_did(new_pairwise_info.pw_did.to_string())
+            .set_service_endpoint(new_service_endpoint)
+            .set_keys(new_recipient_keys, new_routing_keys)
+            .ask_for_ack()
+            .set_thread_id(&request.get_thread_id())
+            .set_out_time();
+
+        sign_connection_response(wallet, &self.pairwise_info.pw_vk, response).await
+    }
+
+    /// Processes a [`Request`] and transitions to [`InviterConnection<Requested>`].
+    ///
+    /// # Errors
+    ///
+    /// Will return an error if either:
+    ///     * the [`Request`]'s thread ID does not match with the expected thread ID from an invitation
+    ///     * the [`Request`]'s DidDoc is not valid
+    ///     * generating new [`PairwiseInfo`] fails
+    pub async fn handle_request<T>(
+        self,
+        wallet: &Arc<dyn BaseWallet>,
+        request: Request,
+        new_service_endpoint: String,
+        new_routing_keys: Vec<String>,
+        transport: &T,
+    ) -> VcxResult<InviterConnection<Requested>>
+    where
+        T: Transport,
+    {
+        trace!(
+            "Connection::process_request >>> request: {:?}, service_endpoint: {}, routing_keys: {:?}",
+            request,
+            new_service_endpoint,
+            new_routing_keys,
+        );
+
+        // There must be some other way to validate the thread ID other than cloning the entire Request
+        verify_thread_id(self.state.thread_id(), &A2AMessage::ConnectionRequest(request.clone()))?;
+
+        // If the request's DidDoc validation fails, we generate and send a ProblemReport.
+        // We then return early with the provided error.
+        if let Err(err) = request.connection.did_doc.validate() {
+            error!("Request DidDoc validation failed! Sending ProblemReport...");
+
+            self.send_problem_report(
+                wallet,
+                &err,
+                &request.get_thread_id(),
+                &request.connection.did_doc,
+                transport,
+            )
+            .await;
+
+            Err(err)?;
+        }
+
+        // Generate new pairwise info that will be used from this point on
+        // and incorporate that into the response.
+        let new_pairwise_info = PairwiseInfo::create(wallet).await?;
+        let signed_response = self
+            .build_response(
+                wallet,
+                &request,
+                &new_pairwise_info,
+                new_service_endpoint,
+                new_routing_keys,
+            )
+            .await?;
+
+        let did_doc = request.connection.did_doc;
+        let state = Requested::new(signed_response, did_doc);
+
+        Ok(Connection {
+            source_id: self.source_id,
+            pairwise_info: new_pairwise_info,
+            initiation_type: self.initiation_type,
+            state,
+        })
+    }
+
+    pub fn get_invitation(&self) -> &Invitation {
```

Comment:
pulling at straws here; but it might be worth adding a doc that if `into_invited` was used to reach this state, and then `get_invitation` is called, the `Invitation` that is returned is meaningless and not representative of anything real

Reply:
Good idea. Unfortunately `into_invited` is a necessary evil for backwards compatibility.

The way I see that this would ideally behave is that you generate an `Inviter`, call `create_invitation`, publish the invitation, and whenever you get a `ConnectionRequest` for that invitation you clone the `Inviter` and continue the protocol from there.

The way the previous state machine was designed was that an `Inviter` would respond to a `ConnectionRequest` in both the `Invited` and `Initial` states. So to kind-of make it consistent across the FFI and agent implementation, there's the `into_invited` method to transition the inviter into `Invited` from `Initial`, so it can then process the `ConnectionRequest`.

I'll also document that using `into_invited()` is discouraged and you should clone the `Connection<Inviter, Invited>` instead.

Reply:
Added more comments.